### ISSUE TITLE
Git commands from the command platte are unavailable during git operations

### ISSUE THREAD

**Issue by johnib** [microsoft/vscode#196699]
Type: <b>Bug</b>

Hi VS Code team,

For years, I've been using VS Code to execute git operations (from the amazing Command Platte).
I always explained this to my colleagues as "you can talk to this platte in plain english, and the right command would pop up before you finish... boosting your speed".

It did boost my performance, I was able to send commands through that Platte so fast, that they sort of got "piped" until the git executable would finish the first command.

However, in recent version of VS Code that ability has disappeared. 
I really have to WAIT for the git operation to finish, for example wait 5s for the pull command, before I can issue the next command, for example checkout to new branch.

The command platte does not show the git commands anymore, while the git executable is running.
I hope you can fix it. It worked amazing for me in the past.

VS Code version: Code 1.83.1 (f1b07bd25dfad64b0167beb15359ae573aecd2cc, 2023-10-10T23:57:32.750Z)
OS version: Darwin arm64 23.0.0
Modes:

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Apple M1 Max (10 x 24)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: enabled_on<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>video_decode: enabled<br>video_encode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled<br>webgpu: enabled|
|Load (avg)|6, 4, 4|
|Memory (System)|32.00GB (1.63GB free)|
|Process Argv|--crash-reporter-id ee249b38-c6d6-4799-b265-ddff17a8fe7f|
|Screen Reader|no|
|VM|0%|
</details><details><summary>Extensions (34)</summary>

Extension|Author (truncated)|Version
---|---|---
vscode-base64|ada|0.1.0
gitignore|cod|0.9.0
vscode-markdownlint|Dav|0.52.0
githistory|don|0.6.20
xml|Dot|2.5.1
json-tools|eri|1.0.2
unixtime-converter|esp|0.0.2
vscode-solution-explorer|fer|0.8.4
pinescript|Jey|3.0.1
json-escaper|jos|1.1.2
guid|kos|0.0.2
rainbow-csv|mec|3.8.0
vscode-azureresourcegroups|ms-|0.7.5
vscode-docker|ms-|1.27.0
vscode-dotnet-runtime|ms-|1.8.1
vscode-kubernetes-tools|ms-|1.3.15
remote-containers|ms-|0.315.1
remote-ssh|ms-|0.106.5
remote-ssh-edit|ms-|0.86.0
azure-account|ms-|0.11.6
powershell|ms-|2023.8.0
remote-explorer|ms-|0.4.1
azurerm-vscode-tools|msa|0.15.11
color-highlight|nau|2.5.0
vscode-yaml|red|1.14.0
r|REd|2.8.2
vscode-sort-json|ric|1.20.0
svg-preview|Sim|2.8.3
timestamp-converter|Sta|0.0.1
vscode-styled-components|sty|1.7.8
azure-event-hub-explorer|Sum|0.1.1
vscode-icons|vsc|12.6.0
highlight-bad-chars|wen|0.0.6
json-scape-unscape|yuv|1.3.0

</details><details>
<summary>A/B Experiments</summary>

```
vsliv368cf:30146710
vsreu685:30147344
python383cf:30185419
vspor879:30202332
vspor708:30202333
vspor363:30204092
vslsvsres303:30308271
vserr242:30382549
pythontb:30283811
vsjup518:30340749
pythonptprofiler:30281270
vsdfh931:30280409
vshan820:30294714
vstes263:30335439
vscoreces:30445986
vscod805cf:30301675
binariesv615:30325510
bridge0708:30335490
bridge0723:30353136
vsaa593:30376534
pythonvs932:30410667
py29gd2263:30856252
vscaat:30438848
vsclangdc:30486549
c4g48928:30535728
dsvsc012cf:30540253
pynewext54:30695312
azure-dev_surveyone:30548225
3biah626:30602489
f6dab269:30613381
a9j8j154:30646983
showlangstatbar:30737416
a2ce3375:30757347
pythonfmttext:30731395
fixshowwlkth:30771522
showindicator:30805244
pythongtdpath:30769146
i26e3531:30792625
pythonnosmt12:30797651
pythonidxpt:30866567
pythonnoceb:30805159
copilotsettingc:30859502
asynctok:30869155
dsvsc013:30795093
dsvsc014:30804076
dsvsc015:30845448
pythontestfixtcf:30871695
pythonregdiag2:30871582
pyreplss1:30865275
pythonmypyd1cf:30859727
pythoncet0:30866520
h48ei257:30870396
pythontbext0:30864172

```

</details>

**Comment by lszomoru** [comment]
@johnib, thank you very much for the feedback. VS Code does not have any git command orchestration mechanism hence each command is being executed immediately when initiated by the user, hence git commands can be executed in parallel.

In the past we have seen cases in which executing git commands in parallel can result in the git repository getting in an unintended state. To avoid this, out of all commands contributed by the git extension we have marked 5 commands as blocking: `checkout`, `commit`, `pull`, `push`, `sync`. When any of these commands are started, users have to wait until the complete before they can execute any other git commands in order.

**Comment by johnib** [comment]
Would you consider to solve this issue with a queuing mechanism? Instead of blocking the user?

Thanks for the response

**Comment by lszomoru** [comment]
At the moment we have no plans for implementing an orchestrator for commands in the git extension. I will leave this issue opened and mark it as "under discussion" so that I can use it to track additional user feedback.

**Comment by johnib** [comment]
Thanks. Last question; is it possible to somehow disable this protection? I admit I've never had issues the way I used it.

**Comment by lszomoru** [comment]
No, there is no way to disable this protection.
